VDA-StorageNode-Registry/contracts/facets/VDAStorageNodeFacet.sol

```diff
@@ -541,4 +545,171 @@ contract VDAStorageNodeFacet is IStorageNode {
 
     emit Slash(nodeDID, reasonCode, distributeTotalAmount, loggerCount, moreInfoUrl);
   }
+
+  /**
+  * @dev see { IStorageNodeRegistry }
+  */
+  function lock(
+    address didAddress,
+    string calldata purpose,
+    uint amount,
+    bool withDeposit,
+    bytes calldata requestSignature,
+    bytes calldata requestProof
+  ) external virtual override {
+    if (bytes(purpose).length == 0) {
+      revert InvalidPurpose();
+    }
+
+    if (amount == 0) {
+      revert InvalidAmount();
+    }
+
+    {
+      bytes memory params = abi.encodePacked(didAddress, purpose, amount, withDeposit);
+      LibVerification.verifyRequest(didAddress, params, requestSignature, requestProof);
+    }
+
+    LibStorageNode.NodeStorage storage ds = LibStorageNode.nodeStorage();
+
+    // Token deposit if needed
+    if (withDeposit) {
+      IERC20(ds.vdaTokenAddress).transferFrom(msg.sender, address(this), amount);
+    } else {
+      int excessAmount = getExcessTokenAmount(didAddress);
+      if (excessAmount < 0 || uint(excessAmount) < amount) {
+        revert InvalidAmount();
+      }
+      ds._stakedTokenAmount[didAddress] = ds._stakedTokenAmount[didAddress] - amount;
+    }
+    ds._didLockPurposeList[didAddress].add(purpose);
+    ds._didLockedAmount[didAddress][purpose] = ds._didLockedAmount[didAddress][purpose] + amount;
+
+    emit Lock(didAddress, purpose, amount);
+  }
+
+  /**
+  * @dev see { IStorageNodeRegistry }
+  */
+  function locked(address didAddress, string calldata purpose) external virtual view override returns(uint) {
+    LibStorageNode.NodeStorage storage ds = LibStorageNode.nodeStorage();
+    return ds._didLockedAmount[didAddress][purpose];
```

Comment:
Perhaps check if locked amount exists and if not, throw a meaningful error message?

Reply:
While creating the `verida-js/node-xx` packages, it might be better to get 0 instead of the transaction reverted.
Once transaction reverted, we should use the `try ... catch ...` statement.
Need to revert error message in the contract?

Reply:
Ok, return `0`